### PR TITLE
Update imgur.py and add auto linking

### PR DESCRIPTION
**Pull request by dngfx**
Add auto-imgur for auto linking of imgur and make the linking a bit fancier.

### CODE REVIEW

File: modules/imgur.py [hunk 1]
```diff
@@ -18,6 +22,95 @@ def _prefix(self, data):
...
+            title = utils.irc.bold(data["title"]) + " " if data["title"] else ""
+            views = data["views"]
+            time = datetime.utcfromtimestamp(data["datetime"]). \
+                strftime("%e %b, %Y at %H:%M")
+            ups = utils.irc.color(str(data["ups"]) + "▲", utils.consts.GREEN)
+            downs = utils.irc.color("▼" + str(data["downs"]), utils.consts.RED)
+            images = data["images_count"]
+            image_plural = "" if images is 1 else "s"
+
+            bracket_left = "(" if title or nsfw else ""
```
Comment (jesopo) [inline modules/imgur.py:72]: `nsfw = "(NSFW)" if nsfw else ""` would be a lot more concise.

File: modules/imgur.py [hunk 2]
```diff
@@ -18,6 +22,95 @@ def _prefix(self, data):
...
 
+    @utils.hook("received.message.channel",
+                priority=EventManager.PRIORITY_LOW)
+    def channel_message(self, event):
+        if not event["channel"].get_setting("auto-imgur", False):
+            return
+
+        self._imgur(event)
+
+    def _imgur(self, event):
```
Comment (jesopo) [inline modules/imgur.py]: why is this it's own function? only 1 thing references it

File: modules/imgur.py [hunk 3]
```diff
@@ -1,14 +1,18 @@
 #--require-config imgur-api-key
 
-import re
-from src import ModuleManager, utils
+import re, json
+from src import ModuleManager, utils, EventManager
+from datetime import datetime
```
Comment (jesopo) [inline modules/imgur.py]: please change to `import datetime` and further down use `datetime.datetime` - to keep in line with the rest of the project

File: modules/imgur.py [hunk 4]
```diff
@@ -18,6 +22,95 @@ def _prefix(self, data):
...
+            reply = self._parse_image(match_image.group(1))
+
+        if not reply:
+            return
+
+        event.eat()
+
+        self.events.on("send.stdout").call(target=event[
+            "channel"], module_name="Imgur", server=event["server"],
+                                           message=reply)
```
Comment (jesopo) [inline modules/imgur.py:49]: wrong indent

File: modules/imgur.py [hunk 5]
```diff
@@ -18,6 +22,95 @@ def _prefix(self, data):
...
+        api_key = self.bot.config["imgur-api-key"]
+        result = utils.http.request(URL_GALLERY % hash,
+            headers={"Authorization": "Client-ID %s" % api_key},
+            json=True)
+
+        if result and result.data["success"]:
+            data = result.data["data"]
+            text = ""
+
+            nsfw = utils.irc.bold("[NSFW] ") if data["nsfw"] else ""
```
Comment (jesopo) [inline modules/imgur.py]: don't needlessly bold a space

File: modules/imgur.py [hunk 6]
```diff
@@ -18,6 +22,95 @@ def _prefix(self, data):
...
+        api_key = self.bot.config["imgur-api-key"]
+        result = utils.http.request(URL_IMAGE % hash,
+            headers={"Authorization": "Client-ID %s" % api_key},
+            json=True)
+
+        if result and result.data["success"]:
+            data = result.data["data"]
+            text = ""
+
+            nsfw = utils.irc.bold("[NSFW] ") if data["nsfw"] else ""
```
Comment (jesopo) [inline modules/imgur.py]: don't needlessly bold a space

File: modules/imgur.py [hunk 7]
```diff
@@ -18,6 +22,95 @@ def _prefix(self, data):
...
+            title = utils.irc.bold(data["title"]) + " " if data["title"] else ""
```
Comment (jesopo) [inline modules/imgur.py]: don't bold the whole topic - use formatting sparingly

File: modules/imgur.py [hunk 8]
```diff
@@ -18,6 +22,95 @@ def _prefix(self, data):
...
+        result = utils.http.request(URL_GALLERY % hash,
+            headers={"Authorization": "Client-ID %s" % api_key},
+            json=True)
+
+        if result and result.data["success"]:
+            data = result.data["data"]
+            text = ""
+
+            nsfw = utils.irc.bold("[NSFW] ") if data["nsfw"] else ""
+            title = utils.irc.bold(data["title"]) + " " if data["title"] else ""
```
Comment (jesopo) [inline modules/imgur.py]: don't bold the whole topic - use formatting sparingly

File: modules/imgur.py [hunk 9]
```diff
@@ -18,6 +22,95 @@ def _prefix(self, data):
...
+            views = data["views"]
+            time = datetime.utcfromtimestamp(data["datetime"]). \
+                strftime("%e %b, %Y at %H:%M")
+            ups = utils.irc.color(str(data["ups"]) + "▲", utils.consts.GREEN)
+            downs = utils.irc.color("▼" + str(data["downs"]), utils.consts.RED)
+            images = data["images_count"]
+            image_plural = "" if images is 1 else "s"
```
Comment (jesopo) [inline modules/imgur.py]: don't use `is`, use `==`

File: modules/imgur.py [hunk 10]
```diff
@@ -18,6 +22,95 @@ def _prefix(self, data):
...
+    def _imgur(self, event):
+        msg = event["message"]
+        reply = ""
+        match_gallery = REGEX_GALLERY.match(msg)
+        match_image = REGEX_IMAGE.match(msg)
+
+        if match_gallery:
+            reply = self._parse_gallery(match_gallery.group(1))
+
+        if match_image and not reply:
```
Comment (jesopo) [inline modules/imgur.py]: instead use `elif match_image:`

File: modules/imgur.py (hunk 1 above)
Comment (dngfx) [reply modules/imgur.py:72]: The ( applies when there's a title too, not just NSFW

File: modules/imgur.py (hunk 2 above)
Comment (dngfx) [reply modules/imgur.py]: Fixed.

File: modules/imgur.py (hunk 1 above)
Comment (jesopo) [reply modules/imgur.py:72]: fair enough.

File: modules/imgur.py [hunk 11]
```diff
@@ -18,6 +23,89 @@ def _prefix(self, data):
...
+        match_gallery = REGEX_GALLERY.match(msg)
+        match_image = REGEX_IMAGE.match(msg)
+
+        if match_gallery:
+            reply = self._parse_gallery(match_gallery.group(1))
+
+        elif match_image:
+            reply = self._parse_image(match_image.group(1))
+
+        if not reply:
```
Comment (jesopo) [inline modules/imgur.py]: change to `else:`

Comment (dngfx) [reply modules/imgur.py]: Shouldn't it be if not reply, as it eats the event if it has a message, otherwise it'll never eat the event?

Comment (jesopo) [reply modules/imgur.py]: no because it'll only hit the `else` if it's not a gallery match and not an image match

Comment (dngfx) [reply modules/imgur.py]: Yep, ignore me, not enough coffee this morning, all fixed.

**Pull request merged**